ir/shared/src/main/scala/org/scalajs/ir/Trees.scala

```diff
@@ -885,7 +885,7 @@ object Trees {
         "extends", "false", "finally", "for", "function", "if", "implements",
         "import", "in", "instanceof", "interface", "let", "new", "null",
         "package", "private", "protected", "public", "return", "static",
-        "super", "switch", "this", "throw", "true", "try", "typeof", "var",
```

Comment:
Removed `"this"` from the reserved word, otherwise get 
> [error] requirement failed: `this` is not a valid global ref name

Reply:
Instead of removing it from this set, consider explicitly allowing it in `isValidJSGlobalRefName`. Because it's still a reserved JS identifier name (that's an ES spec concept), but despite that we allow it as a valid `JSGlobalRef`.

Also we should give a constant like `final val FileLevelThis = "this"`, since it is a bit special compared to regular identifiers.

In `ClassDefChecker`, we should disallow using a `JSGlobalRef(FileLevelThis)` as the `lhs` of an `Assign` node.

Reply:
Thanks!

Later, I think I realized that we needed to define `$fileLevelThis` at the top level of the file. The `js.special.fileLevelThis` should then be transformed into `JSGlobalRef("$fileLevelThis")`, allowing us to reference the global object in any context.


```js
(function(){
'use strict';
var $fileLevelThis = this;
// ...

$xxx.prototype.foo = (function(x) {
  // `val globalObject = js.special.fileLevelThis` should be transformed to
  var $globalObject = $fileLevelThis
  // not `var $globalObject = this` where `this` refers to the instance of `xxx`
  // when we call `foo` via method call `(new xxx()).foo()`.
});
```

Therefore, we didn't need to make `"this"` a special case.

(Oh, we still need to disallow assigning to `JSGlobalRef(FileLevelThis)` as you mentioned)

Reply:
Yes, indeed. That said, we should still put the string `"this"` as the value of the `final val FileLevelThis` (and therefore explicitly allow it). Currently, it prevents a user from accessing the real global variable that happens to be called `$fileLevelThis`. ;)

Reply:
Ah, that's right!